### PR TITLE
New hll++ field type to store HyperLogLogPlusPlus sketches

### PR DESCRIPTION
**Pull request by iverase**
This PR explores the addition of a new field mapper thatch store HyperLogLogPlusPlus(HLL++) sketches. An HLL++ sketch is 
the internal algorithm used to compute cardinalities. 

**Mapper**

The new field is defined in a mapping using the following structure:

```
PUT /example
{
    "mappings": {
        "properties": {
            "hllplusplus": {
                "type": "hll++",
                "precision" : 4
            }
        }
    }
}
```

Where precision must be between 4 and 18.  

`Concerns`: Currently the cardinality aggregation does not expose this precision directly but indirectly via a `precision_threshold` defined as number of documents. I had to add a [mapping](https://github.com/iverase/elasticsearch/blob/22ff0d3617df53a6794ce7f35b8ece53f3d25de0/server/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractHyperLogLog.java#L722) between the `precision_threshold` and `precision` which looks strange.

**HLL input**

```
POST /example/_doc
{
    "hllplusplus" : {
        "hll" :[4, 2, 3, 1, 0, 1, 1, 6, 8, 1, 0, 2, 5, 1, 1, 1]
    }
}
```
where the hll is a fixed length array of bytes where the length is define by the precision (length = 1 << precision). For example for precision 4, the array must have 16 elements but for precision 18, it must have 262144.
 
`Concerns`:  I am unsure if we can handle documents with an array of length 262144?

**Murmur3 input**

```
POST /example/_doc
{
    "hllplusplus" : {
        "murmur3" :[234978346, -186944467]
    }
}
```

where murmur3 is an array of  murmur3 hashes.

`Note`: In order to play nicely with raw data, it needs to be generated the same way we do it internally (e.g murmur3 seed should be set to 0).

**Encoded Murmur3 input**

```
POST /example/_doc
{
    "hllplusplus" : {
        "lc" :[3493, 78945]
    }
}
```

where lc is an array of  encoded murmur3 hashes defined as integers. This is the internal format used by the hll++ algorithm to store murmur3 hashes during the linear counting phase. 

`Note`: The encoding value depends on the precision, e.g the encoded hash might be different for the same value at different precision.

**Aggregations**

This field can be used in standard Cardinality aggregations. It can be used together with standard fields. The precision of the aggregation must be <= the precision of the field.  

Relates #48578

### CODE REVIEW

**Comment by iverase** [comment]
Jenkins run elasticsearch-ci/1

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HllFieldMapper.java [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+            if (subParser != null) {
+                // close the subParser so we advance to the end of the object
+                subParser.close();
+            }
+            context.addIgnoredField(fieldType().name());
+        }
+        context.path().remove();
+    }
+
+    private void writeCompressed(ByteArrayList runLens, ByteBuffersDataOutput dataOutput) throws IOException {
```
Comment (polyfractal) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HllFieldMapper.java:459]: Could we add a comment/javadoc somewhere (maybe here?  Maybe at top with the constants?) about how the different compression scheme work or differ?  Fixed and run-length make sense, but the "custom" confused me :)

Comment (polyfractal) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HllFieldMapper.java:459]: Ah I see, they are mentioned below in the class javadocs.  Ok so maybe this isn't necessary here :)

**Review by polyfractal — COMMENTED** [review]
Had a first pass over it, left a few comments (mostly superficial).  Gonna digest and do a second pass, looking at the datastructures/etc a bit closer.  :+1:

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/support/AnalyticsValuesSourceType.java [hunk 2]
```diff
@@ -42,6 +43,34 @@ public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFa
...
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            throw new AggregationExecutionException("value source of type [" + this.value() + "] is not supported by scripts");
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+            final IndexFieldData<?> indexFieldData = fieldContext.indexFieldData();
+
+            if (!(indexFieldData instanceof IndexHllFieldData)) {
```
[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/support/AnalyticsValuesSourceType.java:67]: I'm surprised checkstyle doesn't complain about this, TBH.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HllFieldMapper.java [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+                @Override
+                public LeafHllFieldData load(LeafReaderContext context) {
+
+                    return new LeafHllFieldData() {
+                        @Override
+                        public HllValues getHllValues() throws IOException {
+                            try {
+                                final BinaryDocValues values = DocValues.getBinary(context.reader(), fieldName);
+                                final ByteArrayDataInput dataInput = new ByteArrayDataInput();
+                                final InternalFixedLengthHllValue fixedValue = new InternalFixedLengthHllValue();
```
Comment (polyfractal) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HllFieldMapper.java:261]: Just for my own knowledge, is the idea here to allocate one of each encoding types, and then switch/reset between them as we encounter different encodings on each segment?

**Review by polyfractal — CHANGES_REQUESTED** [review]
Had another look through.  I think it looks good, no real additional comments about the code/algo.

Could we add some unit tests that verify the merging behavior between different precisions?  Feels like that's where this is most likely to break since that's the newest part.  Probably doesn't need to be a full AggregatorTestCase, just some unit tests exercising the merging part?

Maybe some encoding/decoding tests for the different compressions too?

A thought occurred to me, and I think it's probably a bad idea, but wanted to mention:  I wonder if we should offer a "linear counting" mode for the input format too?  Same reason we use it internally, it's considerably smaller for a few elements.  But that would make the input format even _more_ complicated for clients, and I'm not sure what we'd do if the user sends us more values than they are supposed to (reject?  go ahead and convert?)

Dunno... but thought I'd mention it :)

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HllFieldMapper.java (hunk 3 above)
Comment (iverase) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HllFieldMapper.java:261]: Yes, that is the idea so when you read a value, read the first byte with the encoding type, reset the corresponding object and return it.

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/metrics/HyperLogLog.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+/**
+ * Hyperloglog counter, implemented based on pseudo code from
+ * http://static.googleusercontent.com/media/research.google.com/fr//pubs/archive/40671.pdf and its appendix
+ * https://docs.google.com/document/d/1gyjfMHy43U9OWBXxfaeG-3MjGzejW1dlpyMwEYAAWEI/view?fullscreen
+ *
+ * Trying to understand what this class does without having read the paper is considered adventurous.
+ *
+ * It supports storing several HyperLogLog structures which are identified by a bucket number.
+ */
+public final class HyperLogLog implements Releasable {
```
[the remark this answers is not in the document]
Comment (iverase) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/metrics/HyperLogLog.java]: As we are only adding HLL, it seems a waste of effort to use HLL++ as we would need an extra array for storing the algorithm and continuous checks to make sure the current algorithm is HLL. The interfaces have change a bit so I hope it is more clear now.

**Comment by jpountz** [comment]
> The idea would be they can provide an object with an hll attribute or with a murmur3 attribute. If murmur3 then we would be expecting Linear Counting.

+1

**Pull request closed without merging**